src/expr/core/src/expr/expr_udf.rs

```diff
@@ -440,211 +202,148 @@ impl Build for UserDefinedFunction {
-    // Interval between two successive probes of the UDF DNS.
-    const DNS_PROBE_INTERVAL_SECS: u64 = 5;
-    // Timeout duration for performing an eager DNS resolution.
-    const EAGER_DNS_RESOLVE_TIMEOUT_SECS: u64 = 5;
-    const REQUEST_TIMEOUT_SECS: u64 = 5;
-    const CONNECT_TIMEOUT_SECS: u64 = 5;
-
-    if let Some(s) = addr.strip_prefix("http://") {
-        addr = s;
-    }
-    if let Some(s) = addr.strip_prefix("https://") {
-        addr = s;
-    }
-    let host_addr = addr.parse::<HostAddr>().map_err(|e| {
-        arrow_udf_flight::Error::Service(format!(
-            "invalid address: {}, err: {}",
-            addr,
-            e.as_report()
-        ))
-    })?;
-    let channel = LoadBalancedChannel::builder((host_addr.host.clone(), host_addr.port))
-        .dns_probe_interval(std::time::Duration::from_secs(DNS_PROBE_INTERVAL_SECS))
-        .timeout(Duration::from_secs(REQUEST_TIMEOUT_SECS))
-        .connect_timeout(Duration::from_secs(CONNECT_TIMEOUT_SECS))
-        .resolution_strategy(ResolutionStrategy::Eager {
-            timeout: tokio::time::Duration::from_secs(EAGER_DNS_RESOLVE_TIMEOUT_SECS),
-        })
-        .channel()
-        .await
-        .map_err(|e| {
-            arrow_udf_flight::Error::Service(format!(
-                "failed to create LoadBalancedChannel, address: {}, err: {}",
-                host_addr,
-                e.as_report()
-            ))
-        })?;
-    Ok(channel.into())
-}
-
-/// Get or create a wasm runtime.
-///
-/// Runtimes returned by this function are cached inside for at least 60 seconds.
-/// Later calls with the same binary will reuse the same runtime.
-#[cfg_or_panic(not(madsim))]
-pub fn get_or_create_wasm_runtime(binary: &[u8]) -> Result<Arc<WasmRuntime>> {
-    static RUNTIMES: LazyLock<Cache<md5::Digest, Arc<WasmRuntime>>> = LazyLock::new(|| {
-        Cache::builder()
-            .time_to_idle(Duration::from_secs(60))
-            .build()
-    });
-
-    let md5 = md5::compute(binary);
-    if let Some(runtime) = RUNTIMES.get(&md5) {
-        return Ok(runtime.clone());
-    }
-
-    let runtime = Arc::new(arrow_udf_wasm::Runtime::new(binary)?);
-    RUNTIMES.insert(md5, runtime.clone());
-    Ok(runtime)
-}
-
-/// Returns true if the arrow flight error is caused by a connection error.
-fn is_connection_error(err: &arrow_udf_flight::Error) -> bool {
-    match err {
-        // Connection refused
-        arrow_udf_flight::Error::Tonic(status) if status.code() == tonic::Code::Unavailable => true,
-        _ => false,
-    }
-}
-
-fn is_tonic_error(err: &arrow_udf_flight::Error) -> bool {
-    matches!(
-        err,
-        arrow_udf_flight::Error::Tonic(_)
-            | arrow_udf_flight::Error::Flight(arrow_flight::error::FlightError::Tonic(_))
-    )
+/// Monitor metrics for UDF.
+#[derive(Debug, Clone)]
+struct MetricsVec {
+    /// Number of successful UDF calls.
+    success_count: IntCounterVec,
```

Comment:
Better to use `LabelGuardedIntCounterVec`, etc. https://github.com/risingwavelabs/risingwave/issues/14838

Reply:
let's do it in the following PR.